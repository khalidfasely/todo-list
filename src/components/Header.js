import { connect } from "react-redux";
import { startLogout } from "../actions/auth";

export const Header = ({ startLogout }) => {
    return (
        <header className="header">
            <div className="header__logout-button-container">
                <button
                    className="header__logout-button"
                    data-testid='logout-button'
                    onClick={startLogout}
                >
                    Logout
                </button>
            </div>
            <h1 className="header__title">
                Todo List
            </h1>
        </header>
    );
};

const mapDispatchToProps = (dispatch) => ({
    startLogout: () => dispatch(startLogout())
});

export default connect(undefined, mapDispatchToProps)(Header);