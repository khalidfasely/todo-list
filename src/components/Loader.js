const Loader = () => (
    <div className="loader-page">
        <div className="loader-container">
            <div className="loader">
                <h2 data-text='Loading...'>Loading...</h2>
            </div>    
        </div>
    </div>
);

export default Loader;