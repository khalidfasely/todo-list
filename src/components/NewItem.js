import { useState } from 'react';
import { connect } from 'react-redux';
import moment from 'moment';
import { startAddTodoItem } from '../actions/todoList';

export const NewItem = ({ startAddTodoItem }) => {
    const [ newItem, setNewItem ] = useState('');
    const [ error, setError ] = useState('');
    //const [ todoList, setTodoList ] = useState([]);

    const onNewItemChange = (e) => {
        setNewItem(e.target.value);
    };

    const onAddItemToList = (e) => {
        setError('');
        e.preventDefault();
        const newTodoItem = {
            content: newItem,
            active: true,
            time: moment().format(),
            timeOfDeactive: undefined
        };
        if(newItem.replace(/\s/g, '')) {
            startAddTodoItem(newTodoItem);
            setNewItem('');
        } else {
            setError('Should fill the input!');
        };
    };

    //const onRemoveItem = (index) => {
    //    setTodoList([...todoList.filter((item, indexL) => index !== indexL)]);
    //};

    //const onDeactiveItem = (index) => {
    //    console.log(`Item number ${index} is done.`);
    //};

    return (
        <div className='form-container'>
            <form className='form' data-testid='form' onSubmit={onAddItemToList}>
                <div>
                    <input
                        data-testid='input'
                        placeholder="Add Something To do later!"
                        type='text'
                        maxLength={50}
                        value={newItem}
                        onChange={onNewItemChange}
                    />
                    <abbr title='Add Task ToDo'>
                        <button>+</button>
                    </abbr>
                </div>
                {error && <i className='form__error'>{error}</i>}
            </form>
        </div>
    );
};

const mapDispatchToProps = (dispatch) => ({
    startAddTodoItem: (newTodoItem) => dispatch(startAddTodoItem(newTodoItem))
});

export default connect(undefined, mapDispatchToProps)(NewItem);