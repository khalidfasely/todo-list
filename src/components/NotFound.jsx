import { Link } from "react-router-dom";

const NotFound = () => {
    return (
        <div>404 - Not Found <Link to='/home'>Go home.</Link></div>
    );
};

export default NotFound;