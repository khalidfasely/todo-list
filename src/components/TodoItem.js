import moment from "moment"
import { connect } from "react-redux";
import { startDeactiveTodoItem, startRemoveTodoItem } from "../actions/todoList";

export const TodoItem = ({
        id,
        content,
        time,
        active,
        timeOfDeactive,
        startDeactiveTodoItem,
        startRemoveTodoItem
    }) => {
    
    const onDeactiveTodoItem = () => {
        startDeactiveTodoItem(id);
    };

    const onRemoveTodoItem = () => {
        startRemoveTodoItem(id);
    };

    return (
        <div>
            <div>{content}</div>
            <div>{moment(time).fromNow()}</div>
            {
                active ?
                <button data-testid='done-button' onClick={onDeactiveTodoItem}>Done</button> :
                <abbr title={moment(timeOfDeactive).fromNow()}>
                    <button disabled={true}>Already Done</button>
                </abbr>
            }
            <button data-testid='x-button' onClick={onRemoveTodoItem}>X</button>
        </div>
    )
}

const mapDispatchToProps = (dispatch) => ({
    startRemoveTodoItem: (id) => dispatch(startRemoveTodoItem(id)),
    startDeactiveTodoItem: (id) => dispatch(startDeactiveTodoItem(id))
});

export default connect(undefined, mapDispatchToProps)(TodoItem);