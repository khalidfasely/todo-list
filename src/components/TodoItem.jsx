import moment from "moment"
import { connect } from "react-redux";
import { startDeactiveTodoItem, startRemoveTodoItem } from "../actions/todoList";

export const TodoItem = ({
        id,
        content,
        time,
        active,
        timeOfDeactive,
        startDeactiveTodoItem,
        startRemoveTodoItem
    }) => {
    
    const onDeactiveTodoItem = () => {
        startDeactiveTodoItem(id);
    };

    const onRemoveTodoItem = () => {
        startRemoveTodoItem(id);
    };

    return (
        <div className="item-container">
            <h4 className="item__content">{content}</h4>
            <div className="item__right">
                <div className="item__time">{moment(time).fromNow()}</div>
                {
                    active ?
                    <abbr title="Already done?">
                        <button className="item__button deactive" data-testid='done-button' onClick={onDeactiveTodoItem}>&#x2713;</button>
                    </abbr> :
                    <abbr title={`Done ${moment(timeOfDeactive).fromNow()}`}>
                        <button className="item__button deactivated" disabled={true}>&#x2713;</button>
                    </abbr>
                }
                <abbr title="Remove From List">
                    <button className="item__button remove" data-testid='x-button' onClick={onRemoveTodoItem}>X</button>
                </abbr>
            </div>
        </div>
    )
}

const mapDispatchToProps = (dispatch) => ({
    startRemoveTodoItem: (id) => dispatch(startRemoveTodoItem(id)),
    startDeactiveTodoItem: (id) => dispatch(startDeactiveTodoItem(id))
});

export default connect(undefined, mapDispatchToProps)(TodoItem);