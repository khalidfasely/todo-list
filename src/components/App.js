import Header from "./Header";
import NewItem from "./NewItem";
import '../firebase/firebase';
import { useEffect } from "react";
import database from "../firebase/firebase";

function App() {
  //useEffect(() => {
  //  database.ref('users')
  //  .once('value')
  //  .then((snapshot) => {
  //    const todoListData = [];
  //  
  //    snapshot.forEach((childSnapshot) => {
  //      todoListData.push({
  //        //id: childSnapshot.key,
  //        ...childSnapshot.val()
  //      });
  //    });
  //  
  //    console.log(todoListData);
  //  });
  //}, [])
  return (
    <div className="App">
      <Header />
      <NewItem />
    </div>
  );
};

export default App;
