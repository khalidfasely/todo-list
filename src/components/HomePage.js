import NewItem from "./NewItem";
import Header from "./Header";

const HomePage = () => {
    return (
        <div>
            <Header />
            <NewItem />
        </div>
    )
};

export default HomePage;