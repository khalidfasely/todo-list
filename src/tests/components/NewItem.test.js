import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { NewItem } from '../../components/NewItem';

let startAddTodoItem;

beforeEach(() => {
    startAddTodoItem: jest.fn()
});

test('Should render NewItem component correctly', () => {
    const { asFragment } = render(<NewItem />);
    expect(asFragment()).toMatchSnapshot();
});

test('Should change input value correctly', () => {
    const { getByTestId } = render(<NewItem />);

    const inputEl = getByTestId('input');
    expect(inputEl.value).toBe('');

    fireEvent.change(inputEl, {
        target: {
            value: 'new todo item'
        }
    });
    expect(inputEl.value).toBe('new todo item');
});

//test('Should call startAddTodoItem on form submit', () => {
//    const onStartAddTodoItemSpy = jest.fn();
//    
//    const { getByTestId } = render(<NewItem startAddTodoItem={onStartAddTodoItemSpy} />);
//    
//    const formEl = getByTestId('form');
//    
//    fireEvent.submit(formEl);
//    
//    expect(onStartAddTodoItemSpy).toHaveBeenCalledTimes(0);
//
//    const inputEl = getByTestId('input');
//
//    fireEvent.change(inputEl, {
//        target: {
//            value: 'new todo item'
//        }
//    });
//
//    expect(onStartAddTodoItemSpy).toHaveBeenCalledTimes(1);
//});

test('Should render error on form submit with empty input', () => {
    const { getByTestId } = render(<NewItem />);
    
    const formEl = getByTestId('form');
    
    fireEvent.submit(formEl);

    const errorElement = screen.getByText(/Should fill the input!/i);
    expect(errorElement).toBeInTheDocument();
});