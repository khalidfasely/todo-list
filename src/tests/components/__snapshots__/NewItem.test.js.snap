// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render NewItem component correctly 1`] = `
<DocumentFragment>
  <div
    class="form-container"
  >
    <form
      class="form"
      data-testid="form"
    >
      <div>
        <input
          data-testid="input"
          maxlength="75"
          placeholder="Add Something To do later!"
          type="text"
          value=""
        />
        <abbr
          title="Add Task ToDo"
        >
          <button>
            +
          </button>
        </abbr>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
